src/PropertySuggester/Suggesters/SimpleSuggester.php

```diff
@@ -68,40 +102,96 @@ protected function getSuggestions( array $propertyIds, array $idTuples, $limit,
 		if ( !is_float( $minProbability ) ) {
 			throw new InvalidArgumentException( '$minProbability must be float!' );
 		}
-		if ( !$propertyIds ) {
-			return array();
-		}
-
-		$excludedIds = array_merge( $propertyIds, $this->deprecatedPropertyIds );
-		$count = count( $propertyIds );
 
 		$dbr = $this->lb->getConnection( DB_SLAVE );
-		if ( empty( $idTuples ) ){
-			$condition = 'pid1 IN (' . $dbr->makeList( $propertyIds ) . ')';
-		}
-		else{
-			$condition = $dbr->makeList( $idTuples, LIST_OR );
-		}
+		$count = $this->getPropertyCount( $dbr, $propertyIds );
+
 		$res = $dbr->select(
 			'wbs_propertypairs',
 			array( 'pid' => 'pid2', 'prob' => "sum(probability)/$count" ),
-			array( $condition,
-				   'pid2 NOT IN (' . $dbr->makeList( $excludedIds ) . ')',
-				   'context' => $context ),
+			$this->makeGetSuggestionsConditions( $dbr, $propertyIds, $idTuples, $context ),
 			__METHOD__,
 			array(
 				'GROUP BY' => 'pid2',
 				'ORDER BY' => 'prob DESC',
-				'LIMIT'    => $limit,
+				'LIMIT'	=> $limit,
 				'HAVING'   => 'prob > ' . floatval( $minProbability )
-				)
-			);
+			)
+		);
+
 		$this->lb->reuseConnection( $dbr );
 
 		return $this->buildResult( $res );
 	}
 
 	/**
+	 * @param DatabaseBase $dbr
+	 * @param int[] $propertyIds
+	 *
+	 * @throws RuntimeException
+	 * @return int
+	 */
+	private function getPropertyCount( DatabaseBase $dbr, array $propertyIds ) {
+		$count = count( $propertyIds );
+
+		if ( $count === 0 ) {
+			$cacheKey = $this->cacheKeyPrefix . ':' . 'Suggester:PropertyCount';
+			$count = $this->cache->get( $cacheKey );
+
+			if ( $count === false ) {
+				$res = $dbr->selectRow(
+					'wbs_propertypairs',
+					'count(distinct(pid1)) as count',
+					array(),
+					__METHOD__
+				);
+
+				if ( !$res ) {
+					throw new RuntimeException( 'Failed to obtain property count.' );
+				}
+
+				$count = (int)$res->count;
+				$this->cache->set( $cacheKey, $count, $this->cacheDuration );
+			}
+		}
+
+		return $count;
+	}
+
+	/**
+	 * @param DatabaseBase $dbr
+	 * @param int[] $propertyIds
+	 * @param string[] $idTuples
+	 * @param string $context
+	 *
+	 * @return string[]
+	 */
+	private function makeGetSuggestionsConditions( DatabaseBase $dbr, array $propertyIds,
+		array $idTuples, $context
+	) {
+		$conditions = array(
+			'context' => $context
+		);
+
+		$excludedIds = array_merge( $propertyIds, $this->deprecatedPropertyIds );
```

Comment:
columns are part of the rules: if pid1 exists there will be a property with pid2 with a probability of p.
we only want to select rules that contain existing properties on the left and missing properties on the right side, that's why they are used as NOT IN in one case and IN in the other case.